### DISCUSSION TITLE
트러블슈팅: 라벨이 비어있는 값으로('') 찍히는 문제

### DISCUSSION THREAD

**Discussion by bullmouse** [boostcampaitech7/level2-nlp-generationfornlp-nlp-12-lv3#15] [Show and tell]
# 🤔 분석

Training을 진행하던 도중, 

코드상, trainer.train()을 실행시키게 되면 SFTTrainer 객체를 받은 trainer 내부에 있는 train 메소드를 통해 실행시키게 되는데, SFTTrainer의 경우 다음과 같은 구성으로 이루어져 있었다.

```python
trainer = SFTTrainer(
model=model,
train_dataset=train_dataset_4,
eval_dataset=eval_dataset_4,
data_collator=data_collator,
tokenizer=tokenizer,
compute_metrics=compute_metrics, # -> 이 파라미터를 살펴볼 것!
preprocess_logits_for_metrics=preprocess_logits_for_metrics,
peft_config=peft_config,
args=sft_config,
)
```

여러 가지 파라미터가 존재하는데, 이 중에서 **compute_metrics**에 대한 파라미터가 있는 것을 알 수 있다. 이 파라미터는 모델이 예측한 값과 실제 레이블을 비교해서 평가 지표(metric)을 계산해주는 함수이다. 이 함수는 학습 중 또는 평가 중에 모델의 성능을 평가하는데 사용될 수 있다. 

이 함수를 자세하게 들여다보면, 

```python
# metric 계산 함수
def compute_metrics(evaluation_result):
	logits, labels = evaluation_result
	
	# 토큰화된 레이블 디코딩
	labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
	labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
	labels = list(map(lambda x: x.split("<end_of_turn>")[0].strip(), labels))
	labels = list(map(lambda x: int_output_map[x], labels))
	
	# 소프트맥스 함수를 사용하여 로그트 변환
	probs = torch.nn.functional.softmax(torch.tensor(logits), dim=-1)
	predictions = np.argmax(probs, axis=-1)
	
	# 정확도 계산
	acc = acc_metric.compute(predictions=predictions, references=labels)
	return acc
```

가장 마지막에 labels가 정확도를 계산하는 데 references로 들어가는 것을 알 수 있다. 그런데, 여기서 문제가 발생했다. 최종적으로 labels는 비어있는 리스트 형태였고, 비어있는 리스트가 된 이유를 살펴보기 위해 출력을 해보았다.

```python
def compute_metrics(evaluation_result):
	logits, labels = evaluation_result
	print(labels) # 결과: [2, 2, 2, 2, ...., 2, 2, 2, 2, 2]
```

이런 식으로 ID 값이 전부 bos_token으로 채워진 것을 확인할 수 있었다. 즉, label의 출력이 제대로 되어있지 않았다.
# ❓원인

일단, 가장 먼저 고쳐야 할 부분은 바로 **토큰 그 자체**였다. 기존 모델인 gemma와는 달리 upstage에서 제공하는 SOLAR는 다른 탬플릿과 토큰을 사용하고 있다. 탬플릿이 완전히 다르기 때문에 기존의 compute_metrics를 뜯어고칠 필요가 있었다. 

일단, SOLAR의 토큰 구조는 다음과 같다. bos, eos, unk가 전부 이런 형식으로 되어 있었다.

```python
{'bos_token': '<s>',
 'eos_token': '</s>',
 'unk_token': '<unk>',
 'pad_token': '</s>'}
```

그런데, 다음 코드를 보면 토큰 자체가 잘못 연결되었다는 사실을 알 수 있다. 

```python
labels = list(map(lambda x: x.split("<end_of_turn>")[0].strip(), labels))
```

이 뿐만 아니라, 

```python
response_template = "<start_of_turn>model"

data_collator = DataCollatorForCompletionOnlyLM(
response_template=response_template,
tokenizer=tokenizer,
)

print(data_collator)
```

data_collator 항목을 보면, 다음과 같이 DataCollatorForCompletionOnlyLM을 사용하는 것을 알 수 있는데, 이 DataCollator 함수가 upstage의 SOLAR에서는 전혀 동작하지 않았다. (response_template을 **Assistant:** 로 바꿔서 해당 이후의 정답과 eos만 추출하려고 해도 그랬다.) 사실, 내부 코드를 다 뜯어보면 원인을 파악할 수는 있겠으나, 차라리 그냥 직접 data_collator를 만드는 것이 더 빠를 것 같아서 커스터마이징을 하기로 했다.
# 🎯 해결

그리하여, compute_metrics 함수를 다음과 같이 뜯어고쳤다.

```python
def compute_metrics(evaluation_result):
# 이 부분은 이전과 동일하다.
logits, labels = evaluation_result
labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

'''
(할 필요 없는 부분)
종종 labels에 비어있는 값이 들어있던 경우가 있었다.
근데, 애초에 label은 비어있는 값이 들어가서는 안된다. 비어있는 값이 있다면 그건 잘못 구현한 것이다.
아무튼, 관련 에러가 있었기 때문에 비어있는 경우는 빼는 것으로 대체했지만, 
어차피 개수가 맞지 않기에 또 다른 에러가 나올 것이다.
'''
labels = [label for label in labels if label.strip() != '']


'''
(할 필요 없는 부분)
만약 label에 Assistant:가 있는 경우, 이를 제거하는 코드인데,
label에 Assistant:가 딸려 나오는 것도 이상하기 때문에 구현이 잘못된 경우라고 보는 것이 옳음.
'''
labels = [label.replace("Assistant:\n", "").strip() for label in labels]

# </s>(eos)를 기준으로 split을 진행한 뒤에, 앞부분만을 추출 
# 이것도 만약에 label 자체가 숫자로 나오는게 보장되면 할 필요 X
labels = list(map(lambda x: x.split("</s>")[0].strip(), labels))

'''
extract_answer_from_label 실행
마지막 부분 숫자로만 나오기 때문에, label.split()[-1]을 실행함
'''
labels = list(map(extract_answer_from_label, labels))

# 최종적으로 int_output_map을 통해 변환함
labels = list(map(lambda x: int_output_map[x], labels))
  
# 소프트맥스 함수를 사용하여 로그로 변환
probs = torch.nn.functional.softmax(torch.tensor(logits), dim=-1)
predictions = np.argmax(probs, axis=-1)

# 정확도 계산
acc = acc_metric.compute(predictions=predictions, references=labels)
return acc
```

그리고, collator 부분도 다음과 같이 바꾸었다.

```python
def custom_data_collator(features, tokenizer, response_template):
	for feature in features: 
		# 가장 먼저 각각의 input_ids를 뽑아낸다, 
		# labels의 경우에는 input_ids의 길이와 동일한 길이로 맞춘다.
		input_ids = feature['input_ids']
		labels = [-100] * len(input_ids)
		
		# 디코딩된 텍스트를 출력하여 눈으로 확인해본다.
		decoded_text = tokenizer.decode(input_ids)
		print(f"Decoded Text: {decoded_text}")
		
		# response_template 이후의 응답 부분을 추출한다.
		if response_template in decoded_text:
		
			'''
			response_template 이후의 텍스트를 추출해야 한다.
			이를 위해 먼저 response_template부분이 시작하는 곳의 index를 알아낸다.
			그리고 그 이후 부분을 잘라내어 response_text로 사용한다.
			'''
			start_idx = decoded_text.index(response_template) + len(response_template)
			response_text = decoded_text[start_idx:].strip()
			
			# 응답 끝에 있는 </s>를 제거한다.
			if "</s>" in response_text:
				response_text = response_text.split("</s>")[0].strip()

			# 그러면 최종적으로 깔끔하게 숫자만 나오는 것을 확인할 수 있다.
			print(f"Extracted Response: {response_text}")
			
			# 이제 라벨로 설정할 부분은 다시 토크나이저를 거쳐야 한다.
			response_tokens = tokenizer.encode(response_text, add_special_tokens=False)


			'''
			input_ids 내에서 해당하는 위치에 라벨을 설정한다. 
			start_idx는 디코딩 되었을 때의 위치이기 때문에
			그걸 참고해서 decoded_text[:start_idx]를 뽑아낸다. 
			그러면 정답이 나오기 직전까지의 값이 나온다. 

			그러면 이걸 다시 encode한 뒤, 길이를 구하면 
			인코딩이 되었을 경우 정답 토큰이 몇 번째 idx인지 알 수 있다.
			'''
			start_token_idx = len(tokenizer.encode(decoded_text[:start_idx], add_special_tokens=False))

			# 마지막으로 response_tokens가 위치한 인덱스의 id를 labels로 가져온다.
			for i in range(start_token_idx, start_token_idx + len(response_tokens)):
				labels[i] = input_ids[i]


		'''
		반복문을 돌리면, 결과적으로 labels에 label들이 쌓이게 된다. 
		이를 feature['labels']에 넣으면 최종적으로 완성이 된다. 
		'''
		feature['labels'] = labels

	# 길이를 맞추기 위해 나머지 부분을 패딩 토큰으로 채우고(tokenizer.pad) 이를 리턴한다. 
	return tokenizer.pad(features, return_tensors="pt")
```

**Comment by rasauq1122** [comment]
`<end_of_turn>` 이게 Gemma-like 모델에서는 EOS로 사용되고 있었던 건가요? 전 그냥 모델이 말을 번갈아가며 하는 표시라고 생각하고 있었네요?
